feat: add app skip reconcile annotation to optionally bypass application controller processing (Alpha) (#11728)

mikeshng opened the pull request:
Closes #11728

Introduce a new annotation for Application CR that will allow it to skip the Application controller's reconcile/processing. 
```
metadata:
  annotations:
    argocd.argoproj.io/skip-reconcile: "true"
```
Mainly used for third party integration projects to write status to the Application CR without the status being overwritten by the ArgoCD application controller. See #11728 for more details.

Signed-off-by: Mike Ng <ming@redhat.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
Discussed and recommended by the community in the Contributors Office Hours on 2022 Dec. 15th.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ]  I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
Not applicable, internal usage for third party projects integration.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

**For future reference and prosperity:** 

As requested by the community, documenting the community driven discussions that lead to the decision on why this skip reconcile annotation is required:

The initial presentation and suggestion from the [Open Cluster Management (OCM)](https://open-cluster-management.io/) maintainers is to introduce an optional way for ArgoCD app delivery that uses the hub-spoke pattern or pull model. The ApplicationSet controller will need to be updated to generate OCM [ManifestWork CRs](https://open-cluster-management.io/concepts/manifestwork/) which wraps the Application CRs as payload instead of generating the usual Application CRs. Once the ManifestWorks are created on the hub/primary/multicluster-controlplane cluster, the spoke/managed cluster will pull the payload (Application) down to the managed cluster and the Application will reconcile on the spoke/managed cluster accordingly. This requires the ArgoCD project importing the OCM API and making significant changes to the ApplicationSet API and controller.

The ArgoCD community welcomes the idea of having an optional pull model like behavior but did not like the fact that it requires importing a new API and making significant changes to ApplicationSet. The ArgoCD maintainers suggested to instead create a side project that watches for the ArgoCD Applications and then wrap them in OCM ManifestWorks as payload. This is a great idea that has minimal impact to the ArgoCD code base and the OCM maintainers have created a new [OCM argocd-pull-integration](https://github.com/open-cluster-management-io/argocd-pull-integration) project to do exactly that.

In this model, the Application CRs act as templates and should not be reconciled on the primary/hub/multicluster-controlplane cluster. The OCM maintainers would also like the pulled Application status in the remote/spoke clusters sync back to the original primary/hub cluster Application status. If the Application continues to reconcile, then the status will be overwritten by the Application controller. Therefore, the ArgoCD community suggested adding an annotation to skip the reconcile.

The primary concern is the niche aspect of this annotation. It seems to be mainly focus on this particular third party integration. However, it's discussed that maybe this feature might be useful in the case of debugging, migration, disaster recovery, approval, etc. But those use cases need to be further analyze and should not be recommended yet. Therefore, the community decision is to mark this feature as alpha and clearly state the primary usage (and briefly mention the potential future usage) in the documentation. Another community concern is the visibility of an Application that skipped reconcile in the UI. This is a tricky problem because from the OCM integration perspective, it will be nice if the UI experience is consistent with the existing push like model. The community decided that it's better to make UI changes in the future if this annotation picks up more significant usage out side of the primary intended purpose.

mikeshng commented:
CC @jannfis @iam-veeramalla @jnpacker @xiangjingli

jannfis commented on controller/appcontroller.go:
```diff
@@ -1777,6 +1778,18 @@ func (ctrl *ApplicationController) canProcessApp(obj interface{}) bool {
+		if skipVal, ok := annotations[common.AnnotationKeyAppSkipReconcile]; ok {
+			if skipReconcile, err := strconv.ParseBool(skipVal); err == nil {
+				if skipReconcile {
+					log.Infof("Skipping Application reconcile based on annotation %s", common.AnnotationKeyAppSkipReconcile)
```
If we really want to log here, I think we should set up some context for this log entry - e.g. the application name.

You can use `log.WithField("application", app.QualifiedName()).Infof(...)` for that. 

However - this function could be called quite often, and that might result in some log flood. I would change log level to debug for this message. And also, either remove the below warning log statement as well or transform it to debug.

mikeshng replied:
Done as suggested.

jannfis reviewed (changes requested):
Thanks @mikeshng - it looks good & simple enough to me, with a concern in the comments. I do feel this could need a bit of documentation, though. 

And possibly, we could also add an end-to-end test for it :)

mikeshng commented:
Added documentation and e2e tests as @jannfis suggested.

mikeshng commented:
/assign @crenshaw-dev

/assign @alexmt

I pressed the re-request review button on @jannfis and the bot/github removed you guys as the reviewers.

mikeshng commented:
CC @jessesuen

mikeshng commented:
As suggested in the Contributors Office Hours,
- Updated the documentation and PR with Alpha feature warning.
- Added primary and alternative use cases to the documentation.
- Updated the PR thread with community discussion history on why we should add this skip reconcile annotation option.

jannfis reviewed (approved):
LGTM! Great work, @mikeshng, thanks a lot for this awesome contribution!

@crenshaw-dev @jessesuen @alexmt do you want to take another look? Otherwise, I'll merge it this PR in 48 hours from now.

jannfis commented:
@mikeshng can you please rebase and resolve the conflict, though?

crenshaw-dev reviewed (changes requested):
Adding a change request so I can come back and address one more thing. :-)

crenshaw-dev reviewed (approved):
🚀
